### DISCUSSION TITLE
Question about non-blocking NimBLEScan::start() and continuous scan

### DISCUSSION THREAD

**Discussion by mbariola** [h2zero/NimBLE-Arduino#369] [Q&A] [answered]
Hi,

for an application, I need to always be monitoring a set of BLE devices (identified by their MACs), which move around the ESP32 running the NimBLE instance. The ESP32 is stationary.

I would need to figure out in the most real time possible, when one of such devices is close enough to the ESP32, and I thought that RSSI is a good enough approximation. The ESP32 also does other tasks in the mean time, so I thought that scanning in a non blocking way would be the way to go. But I find it difficult to debug, i.e. I see discontinuity or delays in RSSI updates, and I am not sure, even with active scan on, if I am doing it right. 

What is the intended behavior of NimBLE in this case, ie. continuous nonblocking scan? when does it call the callback for a specific MAC/device? 
Once a while per some throttle timer? as many times as it detects it? as often as the device advertises itself?

Many thanks to anyone who can shed some light on this / can suggest the right way to achieve my needs. As to why Real time is needed, it's a toy for kids of a friend, some kind of Red Light / Green light game. They have a mifit band each and the one which gets within a certain radius of the esp32 scores one point :-)

**Comment by h2zero** [comment]
Hi @mbariola, the continuous scan is intended to simply keep scanning and provide data through the callback provided. There is an internal cache that is configurable and that will filter recently seen devices until the cache is full, at which time it is cleared and the previous devices will be reported again.

For your purposes I suggest not using active scan. That has the effect of delaying scan reports until scan responses are recieved, unless there is data you're interested in from those scan responses of course.

**Reply by mbariola** [reply]
Hi @h2zero , thanks for the answer. Yes you're correct, I just need to "see" those devices and have their RSSI values with a temporal resolution as needed.
So I seem to understand that I should then

first of all filter for the needed MACs possibly even before calling the callback (if there's something like a MAC whitelist)
disable such cache to get the "most real time experience", or clear it every tenth of second or so (could you point me to the calls and configs to do so please? this is what I surmise I should do:
 
NimBLEDevice::whiteListAdd(one call per MAC address of the devices I need to track)
NimBLEScan::setFilterPolicy(BLE_HCI_SCAN_FILT_USE_WL), 
call NimBLEScan::clearDuplicateCache() or maybe NimBLEScan::clearResults() periodically (I could not find how to disable the cache itself. I surmise that as long as the callback is called per each new detection, )
disable active caching, I suppose that RSSI levels will still be available
let NimBLEScan::start(0, myCallback, FALSE or TRUE??) run, and I guess that no code will be executed after this line so I'll have to move everything else to the other core
Did I get it all right? thanks

Ah, an additional couple question. I guess that the scan is reading the advertisement of the two bands, so it's also a matter of how frequently those bands advertise and not much can be done about it. I guess also for battery saving reasons. by using a companion app, I could see the bands advertise once a second.
Would the "realtimeness" of it somehow changed if I were to pair the esp32 with such bands and "interact with them" somehow? maybe asking for their service characteristics? thinking about the "realtimeness" of ble audio and ble controls from such bands (ie controlling the paired smartphone). This is more about the BLE protocol, I know; but it's quite a jumble of conflicting google results. Maybe you have some empirical data or knowledge about it? thanks again

**Reply by mbariola** [reply]
This is the best I could manage, using your whitelist client sketch as basis. 
I can see the band I am using to test around once a second, consistent with the advertising interval of the band itself.
Any suggestions on how to improve it are very appreciated. Also, is there a way somewhere to obtain / read the BLE devices' **Measured power** aka **RSSI at 1mt** ?

```
/*
 * NimBLE_Scan_Whitelist demo
 *
 * Created May 16, 2021
 * Author: h2zero
 */
#include <Arduino.h>
#include <NimBLEDevice.h>
#include <PrintEx.h>
StreamEx mySerial = Serial;

NimBLEScan *pBLEScan;
NimBLEAddress a("e8:f5:7d:2f:ab:b4");

    class MyAdvertisedDeviceCallbacks : public BLEAdvertisedDeviceCallbacks
{
    void onResult(BLEAdvertisedDevice *advertisedDevice)
    {
        mySerial.printf("Found %s with RSSI %d \n", advertisedDevice->getAddress().toString().c_str(), advertisedDevice->getRSSI());
        pBLEScan->clearResults();
    }
};

void setup()
{
    Serial.begin(460800);
    NimBLEDevice::init("");
    NimBLEDevice::whiteListAdd(a);
    Serial.println("Scanning...");
    pBLEScan = NimBLEDevice::getScan();
    pBLEScan->setAdvertisedDeviceCallbacks(new MyAdvertisedDeviceCallbacks());
    pBLEScan->setFilterPolicy(BLE_HCI_SCAN_FILT_USE_WL);
    pBLEScan->setMaxResults(0); // do not store the scan results, use callback only.
}

void loop()
{
    pBLEScan->start(0, false);
    Serial.println("This should never be reached");
    delay(2000);
}
```

**Comment by h2zero** [comment]
This is a good start, there are a few things to incorporate to make this more robust though.

Have a look at https://github.com/h2zero/NimBLE-Arduino/blob/master/examples/NimBLE_Scan_Continuous/NimBLE_Scan_Continuous.ino
For your use, having the call to ` NimBLEDevice::setScanDuplicateCacheSize(10);` would seem most appropriate.

In the callback you should not call `pBLEScan->clearResults();` as it could cause an exception, instead use the setting `pBLEScan->setMaxResults(0);` to not store the results as shown the that example.

In the callback you can retrieve the RSSI from the advertised device using `advertisedDevice->getRSSI()`

**Reply by mbariola** [reply]
Hi @h2zero ,

Thanks for the suggestions, I'll incorporate them right now and report. One thing about the setMaxResults(0) - I already had that as you can see in the sketch, but the detection of it did not seem to work properly. I only received one detection then it went silent. Adding the clearResults() restored the once-per-second (I guess once-per-advertisement)I reporting of the whitelisted band. I did not try in combination of the cache size reduction (will do so now), but could you please explain better what constitutes a "duplicate" and the whole process from detection to putting it in the results list to caching? thanks

Will offer the findings with the modified sketch in a little while

**Comment by mbariola** [comment, marked as answer]
Hi @h2zero , here's the results of the new sketch including the setScanDuplicateCacheSize(10) and the setMaxResults(0) statements  

- If I comment out the clearResults() in the callback, I only get one single reading and nothing else. 
- If I leave it commented, and set the duplicate cache size to 10, or 0, I still only get one single reading
- If I uncomment it, I get the once-per-second reading. 
So apparently either I am making some mistake / don't understand something, or I need to clear or remove the specific scan from the results set. The idea of a possible exception is quite scary though! Maybe you can run this sketch with one device, and see if you have different results than mine

**A heads up:  NimBLEDevice::setScanFilterMode(....) does not appear in the doxygen documentation;**


```
/*
 * NimBLE_Scan_Whitelist demo
 *
 * Created May 16, 2021
 * Author: h2zero
 */
#include <Arduino.h>
#include <NimBLEDevice.h>
#include <PrintEx.h>
StreamEx mySerial = Serial;

NimBLEScan *pBLEScan;
NimBLEAddress a("e8:f5:7d:2f:ab:b4");
uint lastAdv = millis();

    class MyAdvertisedDeviceCallbacks : public BLEAdvertisedDeviceCallbacks
{
    void onResult(BLEAdvertisedDevice *advertisedDevice)
    {
        //Serial.printf("Advertised Device: %s \n", advertisedDevice->toString().c_str());
        /*
         * Here we add the device scanned to the whitelist based on service data but any
         * advertised data can be used for your preffered data.
         */
               mySerial.printf("%d: Found %s with RSSI %d \n", millis() -lastAdv, advertisedDevice->getAddress().toString().c_str(), advertisedDevice->getRSSI());
               lastAdv = millis();
               //pBLEScan->clearResults(); // comment-out 1
    }
};

void setup()
{
    Serial.begin(460800);
    NimBLEDevice::setScanDuplicateCacheSize(10); //comment-out 2
    NimBLEDevice::init("");
    NimBLEDevice::whiteListAdd(a);
    Serial.println("Scanning...");
    pBLEScan = NimBLEDevice::getScan();
    pBLEScan->setAdvertisedDeviceCallbacks(new MyAdvertisedDeviceCallbacks());
    pBLEScan->setFilterPolicy(BLE_HCI_SCAN_FILT_USE_WL);
    pBLEScan->setMaxResults(0); // do not store the scan results, use callback only.
}

void loop()
{
 if(pBLEScan->isScanning() == false)
 {
    pBLEScan->start(0, nullptr, false);
 }
 Serial.println("Looping...");
 delay(2000);
}
```

**Reply by h2zero** [reply]
Sorry for the delayed reply.

Some detail of how this works:
When a device advertises its sends advertisement packets at a specified rate for a specified time, the purpose of the caching is to reduce the amount of this "spam" to only trigger the callbacks at a reasonable rate. The device advertisement rate, can be anywhere from 20ms to 10240ms, so the cache setting has the effect that it will store up to X advertisements before erasing itself and if it gets an advertisement from a device already in the cache it will be dropped. This works well when there are many different devices advertising to control the callback rate. For your case this isn't working as expected due to the whitelist as there is only 1 device stored in the cache (sorry about that, I hadn't considered this factor) so the cache is not being filled/erased, thus dropping the advertisements.

To solve this we can simply turn off the duplicate filter, which can be done by adding an extra parameter to `pBLEScan->setAdvertisedDeviceCallbacks` to request duplicates.

Try changing:
```
pBLEScan->setAdvertisedDeviceCallbacks(new MyAdvertisedDeviceCallbacks());
```

To:
```
pBLEScan->setAdvertisedDeviceCallbacks(new MyAdvertisedDeviceCallbacks(), true); // Set wantDuplicates
```

**Reply by mbariola** [reply]
Hi @h2zero , thanks for the additional explanation. So let's say the cache is a bit like an anti-flood or debounce measure. I understand that there's a lot of layering to the workings of the NimBLE library, to accommodate for all possible usage scenarios. 
I tried your modification, and it works! thanks so much. So, for everyone that may have a similar need as me, here's the whole of the working code, cleaned of commented-out lines and stuff. When this runs, I get an advertisement every second which is the (unmodifiable) advertising rate of the band.

```
#include <Arduino.h>
#include <NimBLEDevice.h>
#include <PrintEx.h> // just for some sprintf in the code, it's accessory
StreamEx mySerial = Serial;
NimBLEScan *pBLEScan;
NimBLEAddress a("e8:f5:7d:2f:ab:b4");
uint lastAdv = millis();

class MyAdvertisedDeviceCallbacks : public BLEAdvertisedDeviceCallbacks
{
    void onResult(BLEAdvertisedDevice *advertisedDevice)
    {
        mySerial.printf("%d: Found %s with RSSI %d \n", millis() -lastAdv, advertisedDevice->getAddress().toString().c_str(), advertisedDevice->getRSSI());
        lastAdv = millis();
    }
};

void setup()
{
    Serial.begin(460800);
    NimBLEDevice::init("");
    NimBLEDevice::whiteListAdd(a);
    Serial.println("Scanning...");
    pBLEScan = NimBLEDevice::getScan();
    pBLEScan->setAdvertisedDeviceCallbacks(new MyAdvertisedDeviceCallbacks(), true);
    pBLEScan->setFilterPolicy(BLE_HCI_SCAN_FILT_USE_WL);
    pBLEScan->setMaxResults(0); // do not store the scan results, use callback only.
}

void loop()
{
    if(pBLEScan->isScanning() == false)
    {
        pBLEScan->start(0, nullptr, false);
    }
    Serial.println("Looping..."); // just a heartbeat & a delay to avoid piling up scan start calls 
    delay(2000);
}

```

Thanks again for walking me through troubleshooting this behavior!



 Unfortunately I cannot mark your reply as "answer" as GH does not let me mark the reply as answer, so I'll mark the post containing the reply as such. Thanks again :-)

**Reply by NorbertRoller** [reply]
Hi, this is exactly what I was trying to do as well. I tried your example and it doesn't compile in 2024 any more.
NimBLE 2.2.0.
Arduino IDE 2.3.4
ESP32 Wroom-32
1. Error Message: no matching function for call to 'NimBLEAddress::NimBLEAddress(const char [18])'
1. Fix: Added the additional parameter "Type" (0 or 1)

2. Error Message: 'class NimBLEScan' has no member named 'setAdvertisedDeviceCallbacks'
This one I couldn't fix as the compiler is correct.  The "class NimBLEScan" doesn't own this member. At least not in the lib I am using.

Any help is appreciated.

**Reply by thekurtovic** [reply]
Have you looked through the migration guide?
https://github.com/h2zero/NimBLE-Arduino/blob/master/docs/1.x_to2.x_migration_guide.md

It changed and you need to do it like this.
```
/** Set the callbacks to call when scan events occur, no duplicates */
    pBLEScan->setScanCallbacks(...);
```

**Reply by NorbertRoller** [reply]
Hi, many thanks for your fast reply. Meanwhile I had experimented with setScanCallbacks already. 
It seems the callback is not executed at all if the white list filter is on
pBLEScan->setScanCallbacks(new MyAdvertisedDeviceCallbacks()); 
The NimBLEAddress a("fe:2b:c1:ce:75:a0", 0); was verified in another sketch. So it is correct for the unit and the unit is in reach as well.

If I disable the whitelist filter setFilterPolicy(BLE_HCI_SCAN_FILT_NO_WL);
I receive lots of Devices and also the one I wanted to single out.

11:42:12.816 -> I NimBLEScan: New advertiser: fe:2b:c1:ce:75:a0
11:42:12.860 -> D NimBLEScanCallbacks: Discovered: Name: , Address: fe:2b:c1:ce:75:a0, serviceUUID: 0x7081
11:42:12.860 -> D NimBLEScanCallbacks: Result: Name: , Address: fe:2b:c1:ce:75:a0, serviceUUID: 0x7081
11:42:12.860 -> D NimBLEScan: erase device: fe:2b:c1:ce:75:a0
 
Any idea what is wrong ?

**Reply by h2zero** [reply]
Try changing this `NimBLEAddress a("fe:2b:c1:ce:75:a0", 0)` to `NimBLEAddress a("fe:2b:c1:ce:75:a0", 1)`

**Reply by NorbertRoller** [reply]
I tried both and the result is unchanged.

**Reply by NorbertRoller** [reply]
This is the code I am using. Maybe I broke something else...
But it is almost identical to what you used. I had to add the nullptr definition.
```
#include <Arduino.h>
#include <NimBLEDevice.h>
NimBLEScan *pBLEScan;
NimBLEAddress a("fe:2b:c1:ce:75:a0", 0); // 0=Public, 1= Random --- BLE_ADDR_PUBLIC (0) BLE_ADDR_RANDOM (1)
uint lastAdv = millis();
#define nullptr NULL

class MyAdvertisedDeviceCallbacks : public BLEAdvertisedDeviceCallbacks
{
    void onDiscovered(BLEAdvertisedDevice *advertisedDevice)
      {
        Serial.println("onDiscovered");
      }

    void onResult(BLEAdvertisedDevice *advertisedDevice)
    {
    char logString[130];
    sprintf(logString,"%d: Found %s with RSSI %d \n", millis() -lastAdv, advertisedDevice->getAddress().toString().c_str(), advertisedDevice->getRSSI());
    Serial.println(logString);
    lastAdv = millis();
    }
};

void onScanEnd(const NimBLEScanResults& results, int reason) {
        Serial.print("Scan Ended; reason = "); Serial.println(reason);
    }

void setup()
{
    Serial.begin(115200);
    NimBLEDevice::init("");
    NimBLEDevice::whiteListAdd(a);
    Serial.println("Scanning...");
    pBLEScan = NimBLEDevice::getScan();
    pBLEScan->setScanCallbacks(new MyAdvertisedDeviceCallbacks());
    pBLEScan->setFilterPolicy(BLE_HCI_SCAN_FILT_USE_WL); // BLE_HCI_SCAN_FILT_USE_WL
    pBLEScan->setMaxResults(0); // do not store the scan results, use callback only.
     // Test only
    if (NimBLEDevice::onWhiteList(a) != 0) {Serial.println("MAC Adresse is on the Whitelist");}
}

void loop()
{
    if(pBLEScan->isScanning() == false)
    {
      Serial.println("isScanning = false");
      pBLEScan->start(0, nullptr, false);
    }
    else
    {
      Serial.println("isScanning = true");
    }

    Serial.println("Looping..."); // just a heartbeat & a delay to avoid piling up scan start calls 
    delay(2000);
}
```

**Reply by h2zero** [reply]
That is old 1.x version code. the scan callbacks have changed, please see the migration guide and current examples and update accordingly.

**Comment by h2zero** [comment]
Awesome! thanks for the confirmation!

**Reply by Zimbu98** [reply]
I am using mbariola's sketch, and it works fine for some MAC's but weirdly not for others. 

If I run his sketch with Whitelist filter commented out (pBLEScan->setFilterPolicy(BLE_HCI_SCAN_FILT_USE_WL);), I can see my target beacon result on the long list of many BLE devices returned by the scan. For example:

23: Found dd:34:02:06:34:17 with RSSI -29

So, the ESP "sees" my beacon with MAC of dd:34:02:06:34:17

But when I turn on the whitelist and filter for this single MAC, I get no results at all.

The only guess I have is that the whitelist is not only filtering for a specific MAC but also for something else as well. But what?

I'm completely stumped on this one. Any ideas?

Edit: Using BLE_HCI_SCAN_FILT_USE_WL_INITA does not help.

My sketch (basically the same as mbariola's):

```
#include <Arduino.h>
#include <NimBLEDevice.h>
//#include <PrintEx.h> // just for some sprintf in the code, it's accessory
//StreamEx mySerial = Serial;
NimBLEScan *pBLEScan;
NimBLEAddress a("dd:34:02:06:34:17");
uint lastAdv = millis();

class MyAdvertisedDeviceCallbacks : public BLEAdvertisedDeviceCallbacks
{
    void onResult(BLEAdvertisedDevice *advertisedDevice)
    {
        Serial.printf("%d: Found %s with RSSI %d \n", millis() -lastAdv, advertisedDevice->getAddress().toString().c_str(), advertisedDevice->getRSSI());
        lastAdv = millis();
    }
};

void setup()
{
    Serial.begin(115200);
    NimBLEDevice::init("");
    NimBLEDevice::whiteListAdd(a);
    Serial.println("Scanning...");
    pBLEScan = NimBLEDevice::getScan();
    pBLEScan->setAdvertisedDeviceCallbacks(new MyAdvertisedDeviceCallbacks(), true);
    pBLEScan->setFilterPolicy(BLE_HCI_SCAN_FILT_USE_WL);
    pBLEScan->setMaxResults(0); // do not store the scan results, use callback only.
}

void loop()
{
    if(pBLEScan->isScanning() == false)
    {
        pBLEScan->start(0, nullptr, false);
    }
    //Serial.println("Looping..."); // just a heartbeat & a delay to avoid piling up scan start calls 
    delay(500);
}


```

**Reply by h2zero** [reply]
Try setting the address type to 1: `NimBLEAddress a("dd:34:02:06:34:17", 1);`

The reason: the address is a random type (2 MS bits are 1).

**Reply by Zimbu98** [reply]
Wow, that did it! So simple. Thanks very much!

**Reply by Zimbu98** [reply]
Quick question: Your instruction fixed my problem, but I don't really know why. Could you point me to some web link that explains something about the meaning of "the address is a random type (2 MS bits are 1)"?

**Reply by h2zero** [reply]
No problem :smile:. glad it's working. Here is a link with some great info about this https://novelbits.io/bluetooth-address-privacy-ble/

**Reply by Zimbu98** [reply]
I was googling like crazy for anything to do with "MS bits" but kept getting many results for some company called MSbit Software! 

Your link really helps me. Thanks.

**Reply by h2zero** [reply]
Lol, yeah that wasn't worded well. That link has a lot of good BLE info if you ever need to look stuff up.

**Comment by Zimbu98** [comment]
Sorry to bother you again. I am stumped again.

I am trying to build a beacon scanner. The user will input the target MAC address, and the scanner sketch will do something when it "sees" the target beacon. I want the target MAC to be placed on the whitelist. OK, so far, so good. mbariola's sketch does that perfectly for public address MACs. 

The problem I am having is that the user will not know if their beacon MAC is a public address or random static address. So if they input a random static MAC address such as dd:34:02:06:34:17, mbariola's code won't work because the MAC will be missing the beacon type indicator: , 1. 

Searching for clues, I looked at your sample sketch "NimBLE_Scan_Whitelist demo" which is able to scan, see, and add a random static address to the whitelist, but I am not sure why it works there.

This works for random static address beacons in Whitelist Demo sketch: 
NimBLEDevice::whiteListAdd(advertisedDevice->getAddress());

But this doesn't work in mbariola sketch: 
NimBLEAddress a("dd:34:02:06:34:17");
...
NimBLEDevice::whiteListAdd(a);

Thanks for any suggestions!

**Reply by h2zero** [reply]
You could test for the 2 most significant bits of the address being `11` and set the address type to random if they are.

This should get the job done:
```
if (*(a.getNative() + 5) && 0xc0) {
  uint64_t addr = a;
  NimBLEDevice::whiteListAdd(NimBLEAddress(addr, 1));
} else {
  NimBLEDevice::whiteListAdd(a);
}
```

**Reply by Zimbu98** [reply]
Hi "hero"!

That works for random static addresses but doesn't work for public addresses. 

I am checking with these two MACs:

dd:34:02:06:34:17 is random static MAC
dd:33:0a:11:18:89 is public MAC

Here's your code that I slightly modified just to see what's happening:

```
#include <Arduino.h>
#include <NimBLEDevice.h>
NimBLEScan *pBLEScan;
NimBLEAddress a("dd:33:0a:11:18:89");
//dd:34:02:06:34:17
//dd:33:0a:11:18:89

uint lastAdv = millis();

class MyAdvertisedDeviceCallbacks : public BLEAdvertisedDeviceCallbacks
{
    void onResult(BLEAdvertisedDevice *advertisedDevice)
    {
        Serial.printf("%d: Found %s with RSSI %d \n", millis() -lastAdv, advertisedDevice->getAddress().toString().c_str(), advertisedDevice->getRSSI());
        lastAdv = millis();
        Serial.printf("address type: %u\n", advertisedDevice->getAddress().getType());
    }
};

void setup()
{
    Serial.begin(115200);
    delay(2000);
    NimBLEDevice::init("");
    //NimBLEDevice::whiteListAdd(a);

    if (*(a.getNative() + 5) && 0xc0) {
    uint64_t addr = a;
    Serial.println("it's a random static mac address");
    NimBLEDevice::whiteListAdd(NimBLEAddress(addr, 1));
    } else {
    Serial.println("it's a public mac address");
    NimBLEDevice::whiteListAdd(a);
    }
  
    Serial.println("Scanning...");
    pBLEScan = NimBLEDevice::getScan();
    pBLEScan->setAdvertisedDeviceCallbacks(new MyAdvertisedDeviceCallbacks(), true);
    pBLEScan->setFilterPolicy(BLE_HCI_SCAN_FILT_USE_WL);
    pBLEScan->setMaxResults(0); // do not store the scan results, use callback only.
}

void loop()
{
    if(pBLEScan->isScanning() == false)
    {
        pBLEScan->start(0, nullptr, false);
    }
    delay(500);
}
```

Using the above code with random static MAC dd:34:02:06:34:17, I get these results in serial:

it's a random static mac address
Scanning...
2627: Found dd:34:02:06:34:17 with RSSI -32 
address type: 1
1003: Found dd:34:02:06:34:17 with RSSI -41 
address type: 1
999: Found dd:34:02:06:34:17 with RSSI -32 
address type: 1

Using the above code with public MAC dd:33:0a:11:18:89, I get these results in serial:

it's a **random static mac address**
Scanning...

(then nothing since it is appending ,1 to a public MAC address which doesn't work)

So it seems like the "if (*(a.getNative() + 5) && 0xc0)" is not identifying a public MAC correctly.

**Reply by h2zero** [reply]
`dd:33:0a:11:18:89`, is a random address by definition, unless the company ID has the two most significant bits of the address set to `1`, in which case this detection method will fail. There is no solution to this situation if the latter is true unfortunately, it's just the rock 'n roll with BLE addresses when used this way. Unless you know the device address **and** type ahead of time, and know that it won't change, it would be quite difficult to achieve your intended outcome.

PS: 
> Hi "hero"!

LOL

**Reply by Zimbu98** [reply]
But if I add the ,1 after the MAC dd:33:0a:11:18:89:

`NimBLEAddress a("dd:33:0a:11:18:89", 1);    // added the ,1 here`

Nothing shows in the serial even when just doing direct add to whitelist without trying to detect anything first.

```
#include <Arduino.h>
#include <NimBLEDevice.h>
NimBLEScan *pBLEScan;
NimBLEAddress a("dd:33:0a:11:18:89", 1);    // added the ,1 here
//dd:34:02:06:34:17
//dd:33:0a:11:18:89

uint lastAdv = millis();

class MyAdvertisedDeviceCallbacks : public BLEAdvertisedDeviceCallbacks
{
    void onResult(BLEAdvertisedDevice *advertisedDevice)
    {
        Serial.printf("%d: Found %s with RSSI %d \n", millis() -lastAdv, advertisedDevice->getAddress().toString().c_str(), advertisedDevice->getRSSI());
        lastAdv = millis();
        Serial.printf("address type: %u\n", advertisedDevice->getAddress().getType());
    }
};

void setup()
{
    Serial.begin(115200);
    delay(2000);
    NimBLEDevice::init("");
    
    NimBLEDevice::whiteListAdd(a);

//    if (*(a.getNative() + 5) && 0xc0) {
//    uint64_t addr = a;
//    Serial.println("it's a random static mac address");
//    NimBLEDevice::whiteListAdd(NimBLEAddress(addr, 1));
//    } else {
//    Serial.println("it's a public mac address");
//    NimBLEDevice::whiteListAdd(a);
//    }
  
    Serial.println("Scanning...");
```

**Reply by h2zero** [reply]
As you stated before the address that you are using here is public (type `0`), so this method would not work, sadly. I cannot think of any solutions to resolve this situation :disappointed:. The only thing I can think of for this is to add the address to the white list twice, with both public and random address types, but this will limit the list to only 6 devices instead of 12.

**Reply by Zimbu98** [reply]
If dd:33:0a:11:18:89 is a public type (type `0`), then why does `if (*(a.getNative() + 5) && 0xc0)` classify it as a random?

By the way, I only need 3 MACs to go on the whitelist, so I can add each MAC both ways and still only fill up 6 of the whitelist spots. That's what I will do, so thanks much for your help, but my curiosity still requires me to ask the above question. lol

**Reply by h2zero** [reply]
I'm not really sure why that address is not working when defined as random. If you print the address type when that device is scanned what value does it have? I had a look at the controller code to see if I was missing something but it performs the same checks to determine random/public (more accurately it uses 0x40 instead of 0xc0).

**Reply by Zimbu98** [reply]
Basically, your `if (*(a.getNative() + 5) && 0xc0)` code snippet is adding **,1** to **both** dd:34:02:06:34:17 **and** dd:33:0a:11:18:89. So your code snippet thinks they are **both** random static MACs, whereas your scanning library thinks dd:33:0a:11:18:89 is a public MAC. The scanning library then takes revenge on us by refusing to show this beacon MAC in the scan results. 

**Proof:**

_Scenario 1:_ 

If I define the MAC at the top of the sketch as `NimBLEAddress a("dd:33:0a:11:18:89");` and then use normal `NimBLEDevice::whiteListAdd(a);` to add it to whitelist, it appears in scanner results plus its type shows as **0**. So the scanning library seems to consider it a **public MAC**. 

(I am using `Serial.printf("address type: %u\n", advertisedDevice->getAddress().getType());` to show the type after it is scanned.)

Serial results: 

Scanning...
2512: Found dd:33:0a:11:18:89 with RSSI -30 
**address type: 0**
555: Found dd:33:0a:11:18:89 with RSSI -30 
address type: 0
556: Found dd:33:0a:11:18:89 with RSSI -29 
address type: 0
552: Found dd:33:0a:11:18:89 with RSSI -27 
address type: 0

_Scenario 2:_

If I define it as a **random static MAC** using `NimBLEAddress a("dd:33:0a:11:18:89, 1");` (added **,1**), then it does NOT appear in the scanner. The scanning library seems unhappy that I have defined it as random static whereas the library thinks it is a public MAC. 

_Scenario 3:_

If I define the MAC at the top of the sketch as `NimBLEAddress a("dd:33:0a:11:18:89");` with **no type number added**, and then use your code snippet to define the type:

```
if (*(a.getNative() + 5) && 0xc0) {
    uint64_t addr = a;
    Serial.println("added MAC to list as a random static mac address");
    NimBLEDevice::whiteListAdd(NimBLEAddress(addr, 1));
    } else {
    Serial.println("added MAC to list as a public mac address");
    NimBLEDevice::whiteListAdd(a);
    }
```
...it gets defined as **1** and again does not appear in the scanner.

**Double check the Whitelist:**

I double checked if it is going onto the whitelist: the answer is yes, it is (see below), but again, it again does not show on the scan results:

```
#include <Arduino.h>
#include <NimBLEDevice.h>
NimBLEScan *pBLEScan;
NimBLEAddress a("dd:33:0a:11:18:89");
//dd:34:02:06:34:17
//dd:33:0a:11:18:89

uint lastAdv = millis();

class MyAdvertisedDeviceCallbacks : public BLEAdvertisedDeviceCallbacks
{
    void onResult(BLEAdvertisedDevice *advertisedDevice)
    {
        Serial.printf("%d: Found %s with RSSI %d \n", millis() -lastAdv, advertisedDevice->getAddress().toString().c_str(), advertisedDevice->getRSSI());
        lastAdv = millis();
        Serial.printf("address type: %u\n", advertisedDevice->getAddress().getType());
    }
};

void setup()
{
    Serial.begin(115200);
    delay(2000);
    NimBLEDevice::init("");
    
    //NimBLEDevice::whiteListAdd(a);

    if (*(a.getNative() + 5) && 0xc0) {
    uint64_t addr = a;
    Serial.println("added MAC to list as a random static mac address");
    NimBLEDevice::whiteListAdd(NimBLEAddress(addr, 1));
    } else {
    Serial.println("added MAC to list as a public mac address");
    NimBLEDevice::whiteListAdd(a);
    }
  
    Serial.println("Scanning...");
    pBLEScan = NimBLEDevice::getScan();
    pBLEScan->setAdvertisedDeviceCallbacks(new MyAdvertisedDeviceCallbacks(), true);
    pBLEScan->setFilterPolicy(BLE_HCI_SCAN_FILT_USE_WL);
    pBLEScan->setMaxResults(0); // do not store the scan results, use callback only.
}

void loop()
{
    if(pBLEScan->isScanning() == false)
    {
        pBLEScan->start(0, nullptr, false);
    }
    delay(500);
    int whiteC = NimBLEDevice::getWhiteListCount();
    Serial.println(whiteC);
}
```

Result in serial: 

added MAC to list as a **random static mac address**
Scanning...
1
1
1
1
1
1
...

**Reply by h2zero** [reply]
My apologies, the snippet ` if (*(a.getNative() + 5) && 0xc0) ` is incorrect, I typed an extra `&` there :disappointed: . 

However that does not change the situation, the result would be the same. If we look at the most significant bits of the address we can see that `dd` == `11011101`, so this would have them both = `1`. This suggests that the address is random, which means the type provided to the whitelist should also be `1`. However I am stumped as to why the scanner thinks otherwise.

**Reply by Zimbu98** [reply]
Yeah, I figured out that && should be &. 

I even tried this modified code snippet: 

```
if(*(a.getNative() + 5) & 0xc0 == 0xc0) {
    uint64_t addr = a;
    Serial.println("added MAC to list as a random static mac address");
    NimBLEDevice::whiteListAdd(NimBLEAddress(addr, 1));
    } else {
    Serial.println("added MAC to list as a public mac address");
    NimBLEDevice::whiteListAdd(a);
    }
```

But it still put classified **both** dd:34:02:06:34:17 and dd:33:0a:11:18:89 as random static. But like you said, **that seems to be correct**. So the issue seems to be on the scanner side, i.e. why is it incorrectly judging the MAC type then throwing some sort of error when the whitelisted MAC's type code doesn't match. 

The only thing I could dig up is this section of ble.h, but it's like a foreign language to me:

```
#define BLE_ADDR_IS_RPA(addr)     (((addr)->type == BLE_ADDR_RANDOM) && \
                                   ((addr)->val[5] & 0xc0) == 0x40)
#define BLE_ADDR_IS_NRPA(addr)    (((addr)->type == BLE_ADDR_RANDOM) && \
                                   ((addr)->val[5] & 0xc0) == 0x00)
#define BLE_ADDR_IS_STATIC(addr)  (((addr)->type == BLE_ADDR_RANDOM) && \
                                   ((addr)->val[5] & 0xc0) == 0xc0)
```

**Reply by Zimbu98** [reply]
> The only thing I can think of for this is to add the address to the white list twice, with both public and random address types, but this will limit the list to only 6 devices instead of 12.

Bad news: it won't let me add the same MAC address twice, even if the address types are different. So, this:

```
NimBLEAddress b1p("dd:33:0a:11:1a:a5", 0); 
NimBLEAddress b1r("dd:33:0a:11:1a:a5", 1);

NimBLEDevice::whiteListAdd(b1p);
NimBLEDevice::whiteListAdd(b1r);
```

Only adds one single MAC to the whitelist. In this case, it only adds the first one (b1p), and then it ignores the second whiteListAdd. If I check the whitelist count while the scanner is running, it is only 1. 

It also doesn't work if I leave off the type code on the first line. For example:

```
NimBLEAddress b1p("dd:33:0a:11:1a:a5"); 
NimBLEAddress b1r("dd:33:0a:11:1a:a5", 1);

NimBLEDevice::whiteListAdd(b1p);
NimBLEDevice::whiteListAdd(b1r);
```

So I think I need to keep digging for the bug that is causing the scanner to misjudge the type code. Then your code snippet suggestion should work fine.

**Reply by Zimbu98** [reply]
My hacky solution: 

Commented out these lines in NimBLEDevice.cpp

```
/**
    if (NimBLEDevice::onWhiteList(address)) {
        return true;
    }
    */
```

Now it allows me to put two versions of each beacon on the whitelist, one version with type code 1 and another version with type code 0 (or no type code and just let it use the default). 

Victory!!!!

**Reply by h2zero** [reply]
Lol, glad you found that, the comparison code needs some work in this regard to take the type into account, I'm just reluctant to as it will likely break some peoples code.

I did a dive back into the core spec to see what it is that was causing the issue and was reminded of the header in the BLE advertisements. The flag that indicates if the address is public or random resides in this header which is then provided as the address type to the scanner from the controller when a device is detected. 

I found a link that describes this more simply than the core document if you're interested: https://medium.com/rtone-iot-security/a-brief-overview-of-bluetooth-low-energy-79be06eab4df

So the code snippet above is invalid anyway for this use, it would only work to confirm the random status of the MAC.

**Reply by Zimbu98** [reply]
If I understand that link correctly, it is saying that the type code for the beacon's MAC is configured in the beacon's broadcast packet. 

So even though you and I (and your code snippet) think that dd:33:0a:11:18:89 is a **random** static MAC type 1, the beacon's broadcast is incorrectly carrying a TxAddr of 0, i.e. a **public** MAC type code. 

Based on that, it makes sense that when I scan and see this beacon, and print the type code, it shows as 0, since the scanner is probably just getting this type 0 info from the beacon's own broadcast packet. (I had incorrectly assumed that the scanner was doing some incorrect analysis of the MAC to figure out the type code).

I will take a look at the raw broadcast from the beacon to confirm this, but I think it is very likely to be true.

Thanks again for all your help. This has been a fascinating learning experience for me!

**Reply by h2zero** [reply]
I wouldn't say that it is broadcasting and incorrect MAC type, it is certainly likely correct for it, even though the bits are set (just means it will never change). The trouble being, for your use case, is there is no automatic way to detect this from the data provided. This is always the issue with hard coding BLE addresses into projects, the type is vital.
